### PR TITLE
Adding nftables functionality for use in EL8

### PR DESCRIPTION
**Pull request by nicholasmhughes**
### What does this PR do?
This PR is the result of attempting to manage nftables on CentOS 8 in a similar fashion to what I had been doing with iptables on previous versions.

### What issues does this PR fix or reference?
N/A

### Previous Behavior
There were some gaps in functionality for current versions of nftables.

### New Behavior
The nftables execution and state modules are now much closer to the iptables modules and work on nftables versions tested with CentOS 8.

### Tests written?
Yes

### Commits signed with GPG?
Yes

### CODE REVIEW

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: salt/states/nftables.py [hunk 1]
```diff
@@ -477,30 +483,42 @@ def delete(name, family='ipv4', **kwargs):
...
         if ignore in kwargs:
             del kwargs[ignore]
 
     if 'table' not in kwargs:
         kwargs['table'] = 'filter'
 
-    res = __salt__['nftables.check_table'](kwargs['table'], family=family)
-    if not res['result']:
+    check_table = __salt__['nftables.check_table'](kwargs['table'], family=family)
+    if not ignore_absence and not check_table['result']:
```
[the remark this answers is not in the document]
Comment (nicholasmhughes) [reply salt/states/nftables.py]: Sure we can!

**Review by garethgreenaway — DISMISSED** [review]

**Comment by nicholasmhughes** [comment]
re-run all

**Comment by nicholasmhughes** [comment]
re-run docs

**Comment by nicholasmhughes** [comment]
re-run full all

File: salt/modules/nftables.py [hunk 2]
```diff
@@ -355,9 +399,48 @@ def get_rules(family="ipv4"):
     return rules
 
 
+def get_rules_json(family="ipv4"):
+    """
+    .. versionadded:: Sodium
```
Comment (Ch3LL) [inline salt/modules/nftables.py:404]: looks like this will need to be updated since it did not make it in sodium

File: salt/modules/nftables.py [hunk 3]
```diff
@@ -355,9 +399,48 @@ def get_rules(family="ipv4"):
...
+    except (KeyError, ValueError):
+        return rules
+
+    return rules
+
+
 def save(filename=None, family="ipv4"):
     """
-    Save the current in-memory rules to disk
+    .. versionchanged:: Sodium
```
Comment (Ch3LL) [inline salt/modules/nftables.py:439]: here as well

File: salt/modules/nftables.py [hunk 4]
```diff
@@ -1064,9 +1156,120 @@ def flush(table="filter", chain="", family="ipv4"):
...
         ret["result"] = True
         ret["comment"] = "Flushed rules {0}".format(comment)
     else:
         ret["comment"] = "Failed to flush rules {0}".format(comment)
     return ret
+
+
+def get_policy(table="filter", chain=None, family="ipv4"):
+    """
+    .. versionadded:: Sodium
```
Comment (Ch3LL) [inline salt/modules/nftables.py:1169]: here as well

File: salt/modules/nftables.py [hunk 5]
```diff
@@ -1064,9 +1156,120 @@ def flush(table="filter", chain="", family="ipv4"):
...
+                and rule.get("chain", {}).get("type") == table
+            ):
+                return rule["chain"]["policy"]
+    except (KeyError, TypeError, ValueError):
+        return None
+
+
+def set_policy(table="filter", chain=None, policy=None, family="ipv4"):
+    """
+    .. versionadded:: Sodium
```
Comment (Ch3LL) [inline salt/modules/nftables.py:1211]: here as well

File: salt/states/nftables.py [hunk 6]
```diff
@@ -135,7 +135,9 @@ def chain_present(
     """
     .. versionadded:: 2014.7.0
 
-    Verify the chain is exist.
+    .. versionchanged:: Sodium
```
Comment (Ch3LL) [inline salt/states/nftables.py:138]: here as well

File: salt/states/nftables.py [hunk 7]
```diff
@@ -454,27 +464,39 @@ def delete(name, family="ipv4", **kwargs):
         return ret
 
 
-def flush(name, family="ipv4", **kwargs):
+def flush(name, family="ipv4", ignore_absence=False, **kwargs):
     """
     .. versionadded:: 2014.7.0
 
+    .. versionchanged:: Sodium
```
Comment (Ch3LL) [inline salt/states/nftables.py:471]: here as well

File: salt/states/nftables.py [hunk 8]
```diff
@@ -512,3 +536,173 @@ def flush(name, family="ipv4", **kwargs):
         ret["result"] = False
         ret["comment"] = "Failed to flush nftables rules"
         return ret
+
+
+def set_policy(name, table="filter", family="ipv4", **kwargs):
+    """
+    .. versionadded:: Sodium
```
Comment (Ch3LL) [inline salt/states/nftables.py:543]: here as well

File: salt/states/nftables.py [hunk 9]
```diff
@@ -512,3 +536,173 @@ def flush(name, family="ipv4", **kwargs):
...
+    else:
+        ret["result"] = False
+        ret["comment"] = "Failed to set nftables default policy"
+
+    return ret
+
+
+def table_present(name, family="ipv4", **kwargs):
+    """
+    .. versionadded:: Sodium
```
Comment (Ch3LL) [inline salt/states/nftables.py:617]: here as well

File: salt/states/nftables.py [hunk 10]
```diff
@@ -512,3 +536,173 @@ def flush(name, family="ipv4", **kwargs):
...
+        ret["comment"] = "Failed to create table {0} for family {1}".format(
+            name, family
+        )
+
+    return ret
+
+
+def table_absent(name, family="ipv4", **kwargs):
+    """
+    .. versionadded:: Sodium
```
Comment (Ch3LL) [inline salt/states/nftables.py:664]: here as well

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: salt/modules/nftables.py (hunk 2 above)
Comment (nicholasmhughes) [reply salt/modules/nftables.py:404]: Can I replace it with `INSERTSOMEFUTUREVERSIONHERE` so I won't have to come back and do it again? 😆

**Comment by nicholasmhughes** [comment]
@Ch3LL requested changes addressed along with new tests that slid in. I can haz review?

File: salt/modules/nftables.py (hunk 2 above)
Comment (Ch3LL) [reply salt/modules/nftables.py]: thats actually a good idea TBH. Every release we search/replace <release name> for <number version> anyways. It wouldn't be such a bad idea to create a standard NEXTRELEASE notation that we could use. Let me propose the idea to the team in a PR.

**Comment by Ch3LL** [comment]
looks like just pre-commit tests that are failing that need to be fixed up.

File: salt/modules/nftables.py (hunk 2 above)
Comment (Ch3LL) [reply salt/modules/nftables.py]: https://github.com/saltstack/salt/pull/58154

**Comment by nicholasmhughes** [comment]
I think that should fix it

**Comment by nicholasmhughes** [comment]
@Ch3LL , it's not immediately obvious to me why lint is failing.

**Comment by Ch3LL** [comment]
@nicholasmhughes go ahead and rebase and they should no longer fail. There was a linter fail on the head of master that should be fixed by now.

**Review by Ch3LL — APPROVED** [review]

**Comment by Ch3LL** [comment]
@garethgreenaway can we get your re-review here as well

**Review by garethgreenaway — APPROVED** [review]

**Pull request merged**